alfresco-health-processor-platform/src/main/java/eu/xenit/alfresco/healthprocessor/processing/ProcessorService.java

```diff
@@ -116,33 +116,38 @@ private void processNodeBatchInTransaction(Set<NodeRef> nodesToProcess, HealthPr
         log.debug("Plugin '{}' will process #{} nodes", plugin.getClass().getCanonicalName(),
                 nodesToProcess.size());
 
-        Set<NodeHealthReport> reports = transactionHelper
-                .inNewTransaction(() -> plugin.process(nodesToProcess), configuration.isReadOnly());
+        Set<NodeHealthReport> pluginReports = transactionHelper
+                .inNewTransaction(() -> Collections.unmodifiableSet(plugin.process(nodesToProcess)), configuration.isReadOnly());
 
-        validateNodeReports(nodesToProcess, reports, plugin);
+        Set<NodeHealthReport> reports = validateNodeReports(nodesToProcess, pluginReports, plugin);
 
         transactionHelper.inNewTransaction(() -> reportsService.processReports(plugin.getClass(), reports), false);
     }
 
-    private void validateNodeReports(Set<NodeRef> nodesToProcess, Set<NodeHealthReport> reports, HealthProcessorPlugin plugin) {
-        if(reports.size() != nodesToProcess.size()) {
-            log.warn("Plugin '{}' returned #{} reports for #{} nodes", plugin.getClass().getCanonicalName(), reports.size(), nodesToProcess.size());
-            Set<NodeRef> reportedNodes = reports.stream().map(NodeHealthReport::getNodeRef).collect(Collectors.toSet());
+    private Set<NodeHealthReport> validateNodeReports(Set<NodeRef> nodesToProcess, Set<NodeHealthReport> reports, HealthProcessorPlugin plugin) {
+        Set<NodeHealthReport> nodeHealthReports = reports;
+        if(nodeHealthReports.size() != nodesToProcess.size()) {
```

Comment:
The validation is only performed when the size of the sets don't match ? What if there is 1 under-reported and 1 over-reported node ? :thinking: 

Besides checking the size, you could do something like:
```
var diffMap = new HashSet<NodeRef>(nodesToProcess);
diffMap.removeAll(reports.stream().map(report::getNodeRef).collect(Collectors.toList));
assertThat(diffMap).isEmpty();
```

Reply:
I was trying to minimize the impact in the likely case, when there is no bug in a plugin.

Going just by the sizes looked like a candidate to catch most issues without needing much additional processing.

Reply:
Those batches should be not HUGE (<100k nodes or so?) so it should not make that big of a perf impact (on something that already runs in the background anyway).

But not _that_ important either, I'll leave it up to you to decide.